### ISSUE TITLE
dnn readNet failed to load my own tf model

### ISSUE THREAD

**Issue by LJXLJXLJX** [opencv/opencv#13110]
I trained a simple model for classification task. I trained my model and generated a .pb file. I loaded the model wtih dnn readNet and it worked well.

Then, I tried to improve my model and introduce batch normalization. Same as before, I generated my .pb file successfully. But when I tried to load it with dnn readNet again, some errors occurred.
```
net=cv2.dnn.readNet(modelPath)
cv2.error: OpenCV(3.4.3) D:\Build\OpenCV\opencv-
3.4.3\modules\dnn\src\tensorflow\tf_importer.cpp:533: error: (-2:Unspecified error) Input 
[Conv/BatchNorm/cond/FusedBatchNorm/Switch_1:1] for node 
[Conv/BatchNorm/cond/FusedBatchNorm] not found in function 
'cv::dnn::experimental_dnn_34_v7::`anonymous-namespace'::TFImporter::getConstBlob'
```

My net defination code is as follow

```
import tensorflow as tf
from tensorflow.contrib import slim
x = tf.placeholder(tf.float32, [None, 32, 32, 1], name='input')
tf.summary.image('input', x)
y_ = tf.placeholder(tf.float32, [None, 37], name='labels')
is_training = tf.placeholder(tf.bool, [], name='is_training')
conv1 = slim.conv2d(
    inputs=x,
    num_outputs=32,
    kernel_size=[5, 5],
    padding="same",
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
pool1 = slim.max_pool2d(inputs=conv1, kernel_size=[2, 2], stride=2)
conv2 = slim.conv2d(
    inputs=pool1,
    num_outputs=64,
    kernel_size=[5, 5],
    padding="same",
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
pool2 = slim.max_pool2d(inputs=conv2, kernel_size=[2, 2], stride=2)
pool2_flat = tf.reshape(pool2, [-1, 8 * 8 * 64])
dense = slim.fully_connected(
    inputs=pool2_flat,
    num_outputs=1024,
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
logits = slim.fully_connected(inputs=dense, num_outputs=37, scope='logits')
predict = tf.arg_max(logits, 1, name='predict')
tf.add_to_collection('result', predict)
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=logits, labels=y_))
tf.summary.scalar('loss', loss)
correct_prediction = tf.equal(tf.argmax(logits, 1), tf.argmax(y_, 1))
with tf.name_scope('accuracy'):
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
tf.summary.scalar('accuracy', accuracy)
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001)
train_step = slim.learning.create_train_op(loss, optimizer)
merged = tf.summary.merge_all()
```

here is my .pb file and log files
https://suall-my.sharepoint.com/:f:/g/personal/46093_officeent_top/EvbmQSncjoJDuS449krtYCYBTA5kbFCdTxpvfDz0eDkpdw?e=OQoMMC

I am sorry to bother.

**Comment by dkurt** [comment]
@LJXLJXLJX, It'd be very helpful if you attached a `.pb` file (at least with randomly initialized weights).

**Comment by LJXLJXLJX** [comment]
> @LJXLJXLJX, It'd be very helpful if you attached a `.pb` file (at least with randomly initialized weights).

Thanks for replying, I have attached the link to my main post,.

**Comment by dkurt** [comment]
@LJXLJXLJX, please try to serialize the model with `is_training: False`.
![screenshot from 2018-11-12 13-02-49](https://user-images.githubusercontent.com/25801568/48340307-5bc80080-e67b-11e8-989e-bf82da1aa2d0.png)

**Comment by LJXLJXLJX** [comment]
@dkurt  I tired to replace the placeholder is_training with a constant False, but I failed. Can you give me some suggestions?
 Followed is my original model freezing code.
```
# coding=UTF-8
import tensorflow as tf
import os.path
import argparse
from tensorflow.python.framework import graph_util
import numpy as np
from tensorflow.core.framework import graph_pb2
import copy

MODEL_DIR = "model"
MODEL_NAME = "mrz32.pb"
testInput = np.ones([1, 32, 32, 1])

if not tf.gfile.Exists(MODEL_DIR): 
    tf.gfile.MakeDirs(MODEL_DIR)

def freeze_graph(model_folder):
    checkpoint = tf.train.get_checkpoint_state(model_folder) 
    input_checkpoint = checkpoint.model_checkpoint_path  
    output_graph = os.path.join(MODEL_DIR, MODEL_NAME)  

    output_node_names = "logits/BiasAdd"  
    saver = tf.train.import_meta_graph(input_checkpoint + '.meta',
                                       clear_devices=True) 

    graph = tf.get_default_graph()  
    input_graph_def = graph.as_graph_def() 
    with tf.Session() as sess:
        saver.restore(sess, input_checkpoint)  

        print("predictions : ", sess.run("logits/BiasAdd:0",
                                         feed_dict={
                                             "input:0": testInput,
                                             "is_training:0":False
                                         }))  

        output_graph_def = graph_util.convert_variables_to_constants( 
            sess,
            input_graph_def ,
            output_node_names.split(",")  
        )
        with tf.gfile.GFile(output_graph, "wb") as f:  
            f.write(output_graph_def.SerializeToString())  
        print("%d ops in the final graph." % len(output_graph_def.node))  

        for op in graph.get_operations():
            print(op.name, op.values())


if __name__ == '__main__':
    freeze_graph('F:/Project/mrzCharRec/MLCharRecongition/checkpoint')

```

**Comment by dkurt** [comment]
@LJXLJXLJX, try the following code to generate graph definition:
```python
import tensorflow as tf
from tensorflow.contrib import slim
x = tf.placeholder(tf.float32, [None, 32, 32, 1], name='input')
tf.summary.image('input', x)
y_ = tf.placeholder(tf.float32, [None, 37], name='labels')
# is_training = tf.placeholder(tf.bool, [], name='is_training')
is_training = False
conv1 = slim.conv2d(
    inputs=x,
    num_outputs=32,
    kernel_size=[5, 5],
    padding="same",
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
pool1 = slim.max_pool2d(inputs=conv1, kernel_size=[2, 2], stride=2)
conv2 = slim.conv2d(
    inputs=pool1,
    num_outputs=64,
    kernel_size=[5, 5],
    padding="same",
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
pool2 = slim.max_pool2d(inputs=conv2, kernel_size=[2, 2], stride=2)
pool2_flat = tf.reshape(pool2, [-1, 8 * 8 * 64])
dense = slim.fully_connected(
    inputs=pool2_flat,
    num_outputs=1024,
    normalizer_fn=slim.batch_norm,
    normalizer_params={'is_training': is_training, 'decay': 0.95},
    activation_fn=tf.nn.relu)
logits = slim.fully_connected(inputs=dense, num_outputs=37, scope='logits')

with tf.Session() as sess:
    graph_def = sess.graph.as_graph_def()
    with tf.gfile.FastGFile('test.pb', 'wb') as f:
        f.write(graph_def.SerializeToString())
```
Then use it with your checkpoint using `/path/to/tensorflow/tensorflow/python/tools/freeze_graph.py` script

**Comment by LJXLJXLJX** [comment]
@dkurt My problem is perfectly solved, thanks for your effort! But I still have a question. I define my last fc layer with namescope "logits"
`logits = slim.fully_connected(inputs=dense, num_outputs=37, scope='logits')`
But when I try to freeze the graph, it tell me there is no "logits" in the graph,  so I replace it with "logits/BiasAdd"  and it works well.
I can't understand what happend.

**Comment by dkurt** [comment]
@LJXLJXLJX, the thing is that `logits` is a namescope contains two operations: matrix multiplication and bias addition. All at once we call as a fully connected layer.
![screenshot from 2018-11-13 09-15-34](https://user-images.githubusercontent.com/25801568/48394452-b1a4b300-e724-11e8-9d6b-1b8e4bb34fba.png)

Can we close an issue?
